ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSfilemanip.cpp

```diff
@@ -107,92 +114,454 @@ void ini_section_delete(std::string section)
+                    }
+                    RemoveDirectory(FileName);
+                    _tcscpy(FileName,DirPath);
+                }
+                else
+                {
+                    if (FindFileData.dwFileAttributes &
+                        FILE_ATTRIBUTE_READONLY)
+                        _chmod(FileName, _S_IWRITE);
+
+                    if (!DeleteFile(FileName))
+                    {
+                        FindClose(hFind);
+                        return 0;
+                    }
+                    _tcscpy(FileName,DirPath);
+                }
+            }
+            else
+            {
+                if (GetLastError() == ERROR_NO_MORE_FILES)
+                    bSearch = false;
+                else
+                {
+                    FindClose(hFind);
+                    return 0;
+                }
+            }
+        }
+        FindClose(hFind);
+
+        return (RemoveDirectory((char *)str.c_str()) == true);
+    }
+
+    return 0;
+}
+
+int directory_rename(std::string oldname,std::string newname)
+{
+    std::string str1(oldname);
+    std::string str2(newname);
+    std::replace(str1.begin(),str1.end(),'/','\\');
+    std::replace(str2.begin(),str2.end(),'/','\\');
+
+    char rpath1[MAX_PATH];
+    char rpath2[MAX_PATH];
+    GetFullPathName((char *)str1.c_str(),MAX_PATH,rpath1,NULL);
+    GetFullPathName((char *)str2.c_str(),MAX_PATH,rpath2,NULL);
+    str1 = std::string(rpath1);
+    str2 = std::string(rpath2);
+
+    if (!str1.empty())
+    {
+        while (*str1.rbegin() == '\\')
+        {
+            str1.erase(str1.size()-1);
+        }
+    }
+    if (!str2.empty())
+    {
+        while (*str2.rbegin() == '\\')
+        {
+            str2.erase(str2.size()-1);
+        }
+    }
+
+    size_t slash1;
+    size_t slash2;
+    std::string dir1(str1);
+    std::string dir2(str2);
+
+    if (dir1.find('\\') != std::string::npos)
+        slash1 = dir1.find_last_of("\\");
+    else
+    {
+        dir1 = std::string("");
+        slash1 = 0;
+    }
+
+    if (dir2.find('\\') != std::string::npos)
+        slash2 = dir2.find_last_of("\\");
+    else
+    {
+        dir2 = std::string("");
+        slash2 = 0;
     }
+
+    int proceed1 = (str1 != str2.substr(0,str1.length()));
+    int proceed2 = (dir1.substr(0,slash1) == dir2.substr(0,slash2));
+
+    if (proceed1 || (proceed2 && basename((char *)str1.c_str()) != basename((char *)str2.c_str())))
+    {
+        struct stat sb1;
+        struct stat sb2;
+        if (stat((char *)str1.c_str(),&sb1) == 0 &&
+            S_ISDIR(sb1.st_mode) &&
+            stat((char *)str2.c_str(),&sb2) != 0)
+        {
+            return (rename((char *)str1.c_str(),(char *)str2.c_str()) == 0);
+        }
+    }
+
+    return 0;
 }
 
-int directory_delete(std::string dname) {
-  BOOL result = RemoveDirectory(dname.c_str());
-  if (result) return 1;
-  else return 0;
+int directory_copy(std::string dname,std::string newname)
```

Comment:
Windows spares us this complicated piece of logic (and other complicated pieces of logic in this file) via the [`IFileOperation`](https://msdn.microsoft.com/en-us/library/bb775771.aspx) interface. Anything in this file that you can't do simply using a single Win32 call, I recommend doing using that. If we want to support older versions of Windows, you may want to try using [`SHFileOperation`](https://msdn.microsoft.com/en-us/library/bb762164.aspx) instead. I could get behind using a macro test on the current header version to decide which one to use.

Reply:
lol once upon a time my Windows code did use SHFileOperation. So yeah it won't be too hard to add that back in there fortunately.